### ISSUE TITLE
Scrcpy interaction with OPPO Phone is not working

### ISSUE THREAD

**Issue by EloZ27** [Genymobile/scrcpy#1518]
- [ ] I have read the [FAQ](https://github.com/Genymobile/scrcpy/blob/master/FAQ.md).
 - [ ] I have searched in existing [issues](https://github.com/Genymobile/scrcpy/issues).

**Environment**
 - OS: [e.g. Debian, Windows, macOS...]
 - scrcpy version: [e.g. 1.12.1]
 - installation method: [e.g. manual build, apt, snap, brew, Windows release...]
 - device model:
 - Android version: [e.g. 10]

**Describe the bug**

On errors, please provide the output of the console (and `adb logcat` if relevant).

```
Please paste terminal output in a code block.
```

Please do not post screenshots of your terminal, just post the content as text instead.

**Comment by rom1v** [comment]
https://github.com/Genymobile/scrcpy/blob/master/FAQ.md#mouse-and-keyboard-do-not-work

?

**Comment by EloZ27** [comment]
![Screenshot_2020-06-18-02-20-10-96](https://user-images.githubusercontent.com/67069658/84934714-4eea7900-b10a-11ea-9a56-408abd2665d7.png)


Yes, I have already read that. But based in my phone I don't have such option for that. I even looked entirely in my phone, but I haven't found any option for it.

**Comment by rom1v** [comment]
Any hint in `adb logcat`?

What happens if you focus some text area on your device, and execute `adb shell input text hello` from your computer?

**Comment by EloZ27** [comment]
Cant seem to find any hint in the logcat.

adb shell input text hello worked

**Comment by rom1v** [comment]
> adb shell input text hello worked

Oh, that's very weird, then. Pressing `h` `e` `l` `l` `o` in the scrcpy windows does not work?

What if you start with `scrcpy --prefer-text`?

**Comment by EloZ27** [comment]
![123123123123](https://user-images.githubusercontent.com/67069658/85017483-201de280-b19e-11ea-8ce1-a897297f2b17.png)

Hello. Here is the result for the scrcpy --prefer-text.

**Comment by rom1v** [comment]
But what happens when you enter text in the scrcpy window?

**Comment by EloZ27** [comment]
Hmmm. Somehow, typing in the keyboard worked in scrcpy. But the mouse still does not interact

**Comment by rom1v** [comment]
Probably similar to #1347.

Does this work: https://github.com/Genymobile/scrcpy/issues/1347#issuecomment-624039383

**Comment by EloZ27** [comment]
Yea. It some how tapped my screen.

**Comment by rom1v** [comment]
Please test with older versions of scrcpy, if you find one which work, it will help a lot to find the cause.

**Comment by EloZ27** [comment]
I have tried versions down to v1.4, scroll wheel and left mouse button does not work. Middle click and right click worked as well as the keyboard.

**Comment by GetCurious** [comment]
Same phone (F11 Pro), same situation, most keyboard, mid click & right click works **except left click.**

**Comment by rom1v** [comment]
> adb shell input tap 179 179

Does it work or not?

**Comment by GetCurious** [comment]
yup the command works. but left click doesnt

**Comment by rom1v** [comment]
Oh, could you test with that change, please?

```diff
```

**Comment by rom1v** [comment]
Oh, could you test with this change, please:
```diff
diff --git a/app/src/input_manager.c b/app/src/input_manager.c
index 9c22ee0a..7d500efe 100644
--- a/app/src/input_manager.c
+++ b/app/src/input_manager.c
@@ -511,7 +511,8 @@ convert_mouse_button(const SDL_MouseButtonEvent *from, struct screen *screen,
     to->inject_touch_event.position.screen_size = screen->frame_size;
     to->inject_touch_event.position.point =
         screen_convert_window_to_frame_coords(screen, from->x, from->y);
-    to->inject_touch_event.pressure = 1.f;
+    to->inject_touch_event.pressure =
+        from->type == SDL_MOUSEBUTTONDOWN ? 1.f : 0.f;
     to->inject_touch_event.buttons =
         convert_mouse_buttons(SDL_BUTTON(from->button));
 
```

**Comment by GetCurious** [comment]
no difference

**Comment by rom1v** [comment]
> no difference.

Could you please test this change instead:
```diff
diff --git a/server/src/main/java/com/genymobile/scrcpy/Controller.java b/server/src/main/java/com/genymobile/scrcpy/Controller.java
index 71e7ec9c..1f249225 100644
--- a/server/src/main/java/com/genymobile/scrcpy/Controller.java
+++ b/server/src/main/java/com/genymobile/scrcpy/Controller.java
@@ -11,7 +11,7 @@ import java.io.IOException;
 
 public class Controller {
 
-    private static final int DEVICE_ID_VIRTUAL = -1;
+    private static final int DEVICE_ID_VIRTUAL = 0;
 
     private final Device device;
     private final DesktopConnection connection;
```

> can i have your full input_manager.c?

(I use the one from current `dev` branch)

Note: `adb shell input tap ...` calls this: https://github.com/aosp-mirror/platform_frameworks_base/blob/22e3e74e4b3b989d20bb17ca7d54f95b67d6c02c/cmds/input/src/com/android/commands/input/Input.java#L221-L226

**Comment by GetCurious** [comment]
i got nothing (instead and in-addition).

**Comment by rom1v** [comment]
@GetCurious OK, just to be sure, you are building the server (you don't use the prebuilt-server)?

What we have to understand is why this works (`adb shell input tap ...`): https://github.com/aosp-mirror/platform_frameworks_base/blob/22e3e74e4b3b989d20bb17ca7d54f95b67d6c02c/cmds/input/src/com/android/commands/input/Input.java#L221-L226

but this does not:
https://github.com/Genymobile/scrcpy/blob/5fa46ad0c71169f9369f9f3bde9cce3d29ed4b88/server/src/main/java/com/genymobile/scrcpy/Controller.java#L198-L200

**Comment by GetCurious** [comment]
```
$ cd scrcpy
$ vim app/src/input_manager.c
$ vim server/src/main/java/com/genymobile/scrcpy/Controller.java
$ meson x --buildtype release --strip -Db_lto=true
$ ninja -Cx 
$ ./run x
```
anything i miss?

**Comment by rom1v** [comment]
Let's try to use a device id matching the input source, like they do:

```diff
diff --git a/server/src/main/java/com/genymobile/scrcpy/Controller.java b/server/src/main/java/com/genymobile/scrcpy/Controller.java
index 71e7ec9c..c1949f36 100644
--- a/server/src/main/java/com/genymobile/scrcpy/Controller.java
+++ b/server/src/main/java/com/genymobile/scrcpy/Controller.java
@@ -195,12 +195,26 @@ public class Controller {
             }
         }
 
+        int deviceId = getInputDeviceId(InputDevice.SOURCE_TOUCHSCREEN);
+        Ln.i("device id = " + deviceId);
         MotionEvent event = MotionEvent
-                .obtain(lastTouchDown, now, action, pointerCount, pointerProperties, pointerCoords, 0, buttons, 1f, 1f, DEVICE_ID_VIRTUAL, 0,
+                .obtain(lastTouchDown, now, action, pointerCount, pointerProperties, pointerCoords, 0, buttons, 1f, 1f, deviceId, 0,
                         InputDevice.SOURCE_TOUCHSCREEN, 0);
         return device.injectEvent(event);
     }
 
+    private static int getInputDeviceId(int inputSource) {
+        final int DEFAULT_DEVICE_ID = 0;
+        int[] devIds = InputDevice.getDeviceIds();
+        for (int devId : devIds) {
+            InputDevice inputDev = InputDevice.getDevice(devId);
+            if (inputDev.supportsSource(inputSource)) {
+                return devId;
+            }
+        }
+        return DEFAULT_DEVICE_ID;
+    }
+
     private boolean injectScroll(Position position, int hScroll, int vScroll) {
         long now = SystemClock.uptimeMillis();
         Point point = device.getPhysicalPoint(position);
```

**Comment by GetCurious** [comment]
no difference

**Comment by rom1v** [comment]
What's the id printed in the console when you click with this change?

**Comment by GetCurious** [comment]
how do i do that?

**Comment by rom1v** [comment]
When you execute `./run x`, what is the output in the console?

**Comment by GetCurious** [comment]
```
➜  scrcpy git:(dev) ✗ ./run x                                                      
INFO: scrcpy 1.14 <https://github.com/Genymobile/scrcpy>
x/server/scrcpy-server: 1 file pushed, 0 skipped. 107.0 MB/s (33374 bytes in 0.000s)
[server] INFO: Device: OPPO CPH1969 (Android 10)
INFO: Renderer: opengl
INFO: OpenGL version: 4.6.0 NVIDIA 440.100
INFO: Trilinear filtering enabled
INFO: Initial texture: 1080x2336
INFO: New texture: 888x1920
```

**Comment by rom1v** [comment]
And nothing more, even when you click in the scrcpy window?

**Comment by GetCurious** [comment]
that's right... nothing more

**Comment by rom1v** [comment]
With that change: https://github.com/Genymobile/scrcpy/issues/1518#issuecomment-655440431 it should, since a log is added on every click.

Unless the client does not send the events to the device. Let's add more logs:

```diff
diff --git a/app/src/input_manager.c b/app/src/input_manager.c
index 7d500efe..fc0edb6d 100644
--- a/app/src/input_manager.c
+++ b/app/src/input_manager.c
@@ -523,6 +523,14 @@ void
 input_manager_process_mouse_button(struct input_manager *im,
                                    const SDL_MouseButtonEvent *event,
                                    bool control) {
+    LOGI("mouse button: type=%d, which=%d, button=%d, state=%d, clicks=%d, x=%d, y=%d",
+            (int) event->type,
+            (int) event->which,
+            (int) event->button,
+            (int) event->state,
+            (int) event->clicks,
+            (int) event->x,
+            (int) event->y);
     if (event->which == SDL_TOUCH_MOUSEID) {
         // simulated from touch events, so it's a duplicate
         return;
@@ -559,9 +567,12 @@ input_manager_process_mouse_button(struct input_manager *im,
 
     struct control_msg msg;
     if (convert_mouse_button(event, im->screen, &msg)) {
+        LOGI("==== convert_mouse_button OK");
         if (!controller_push_msg(im->controller, &msg)) {
             LOGW("Could not request 'inject mouse button event'");
         }
+    } else {
+        LOGW("==== convert_mouse_button KO");
     }
 }
 
```

**Comment by GetCurious** [comment]
```INFO: mouse button: type=1026, which=0, button=1, state=0, clicks=1, x=323, y=780
INFO: ==== convert_mouse_button OK
```

**Comment by rom1v** [comment]
OK, so the events are correctly forwarded by the client.

One more diff:
```diff
diff --git a/server/src/main/java/com/genymobile/scrcpy/Controller.java b/server/src/main/java/com/genymobile/scrcpy/Controller.java
index 71e7ec9c..1a5fad7f 100644
--- a/server/src/main/java/com/genymobile/scrcpy/Controller.java
+++ b/server/src/main/java/com/genymobile/scrcpy/Controller.java
@@ -71,6 +71,7 @@ public class Controller {
 
     private void handleEvent() throws IOException {
         ControlMessage msg = connection.receiveControlMessage();
+        Ln.i("Received msg type = " + msg.getType());
         switch (msg.getType()) {
             case ControlMessage.TYPE_INJECT_KEYCODE:
                 if (device.supportsInputEvents()) {
@@ -163,6 +164,7 @@ public class Controller {
 
     private boolean injectTouch(int action, long pointerId, Position position, float pressure, int buttons) {
         long now = SystemClock.uptimeMillis();
+        Ln.i("=== injectTouch");
 
         Point point = device.getPhysicalPoint(position);
         if (point == null) {
@@ -170,6 +172,8 @@ public class Controller {
             return false;
         }
 
+        Ln.i("Point = " + point);
+
         int pointerIndex = pointersState.getPointerIndex(pointerId);
         if (pointerIndex == -1) {
             Ln.w("Too many pointers for touch event");

```

**Comment by rom1v** [comment]
https://github.com/Genymobile/scrcpy/issues/1518#issuecomment-655499111

> But probably not related.

Oh in fact, that might be related: the client sends clicks for a window having different dimensions as the ones provided by the device.

If you remove these lines, it should "work" (it may click at a wrong location though):

https://github.com/Genymobile/scrcpy/blob/e99b896ae2b6a59d5f7f2b76a3dd9c29f4838013/server/src/main/java/com/genymobile/scrcpy/Device.java#L134-L138

>     INFO: Initial texture: 1080x2336
>     INFO: New texture: 888x1920

What if you run `scrcpy -m1920`? Do clicks work in that case?

**Comment by GetCurious** [comment]
what the... 
`scrcpy -m1920` worked

**Comment by rom1v** [comment]
One more test, please (with just `./run x`, without `-m` parameter):

```diff
diff --git a/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java b/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java
index d722388c..41afd35b 100644
--- a/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java
+++ b/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java
@@ -82,6 +82,9 @@ public class ScreenEncoder implements Device.RotationListener {
 
                 setSize(format, videoRect.width(), videoRect.height());
                 configure(codec, format);
+                MediaFormat actualFormat = codec.getInputFormat();
+                Ln.i("width = " + actualFormat.getInteger(MediaFormat.KEY_WIDTH));
+                Ln.i("height = " + actualFormat.getInteger(MediaFormat.KEY_HEIGHT));
                 Surface surface = codec.createInputSurface();
                 setDisplaySurface(display, surface, videoRotation, contentRect, unlockedVideoRect, layerStack);
                 codec.start();
```

**Comment by GetCurious** [comment]
```
[server] INFO: width = 2336
[server] INFO: height = 1080
```

**Comment by rom1v** [comment]
OK, and with 
```
MediaFormat actualFormat = codec.getOutputFormat();
```
instead of
```
MediaFormat actualFormat = codec.getInputFormat();
```
?

**Comment by GetCurious** [comment]
```
INFO: scrcpy 1.14 <https://github.com/Genymobile/scrcpy>
...
INFO: Initial texture: 1080x2336
[server] INFO: width = 1080
[server] INFO: height = 2336
INFO: New texture: 888x1920
```

**Comment by rom1v** [comment]
OK, thank you for the tests.

So on the device, we request a specific size (1080x2336), the codec decides to encode in 888x1920 instead (probably because it does not support higher definition), but we can't know what size it decided without decoding the frames :confused:

**Comment by rom1v** [comment]
Just in case, try to move it after the codec is started:

```diff
diff --git a/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java b/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java
index d722388c..3ee09aad 100644
--- a/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java
+++ b/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java
@@ -104,6 +104,10 @@ public class ScreenEncoder implements Device.RotationListener {
         boolean eof = false;
         MediaCodec.BufferInfo bufferInfo = new MediaCodec.BufferInfo();
 
+        MediaFormat actualFormat = codec.getOutputFormat();
+        Ln.i("width = " + actualFormat.getInteger(MediaFormat.KEY_WIDTH));
+        Ln.i("height = " + actualFormat.getInteger(MediaFormat.KEY_HEIGHT));
+
         while (!consumeRotationChange() && !eof) {
             int outputBufferId = codec.dequeueOutputBuffer(bufferInfo, -1);
             eof = (bufferInfo.flags & MediaCodec.BUFFER_FLAG_END_OF_STREAM) != 0;
```

**Comment by GetCurious** [comment]
```
INFO: scrcpy 1.14 <https://github.com/Genymobile/scrcpy>
x/server/scrcpy-server: 1 file pushed, 0 skipped. 159.5 MB/s (33294 bytes in 0.000s)
[server] INFO: Device: OPPO CPH1969 (Android 10)
[server] INFO: width = 1080
[server] INFO: height = 2336


[server] INFO: encode width = 1080
[server] INFO: encode height = 2336


INFO: OpenGL shaders: ENABLED
INFO: Created renderer: opengl
INFO: Renderer: opengl
INFO: OpenGL version: 4.6.0 NVIDIA 440.100
INFO: Trilinear filtering enabled
INFO: Initial texture: 1080x2336
INFO: New texture: 888x1920
```

**Comment by swamikamal** [comment]
i am also getting  same problem. i have oppo f1. I HAVE tried all different version of scrcpy but nothing worked. help me if you are able to correct it

**Comment by rom1v** [comment]
@swamikamal I just gave you the solution: https://github.com/Genymobile/scrcpy/issues/1645#issuecomment-670433990

**Comment by eku** [comment]
May I join this party. I have a Xioami with Adnroid 7.1. and the tip from #1645 does not help.

**Comment by rom1v** [comment]
@eku whtat is the full output in your console when you run scrcpy?

**Comment by eku** [comment]
@rom1v 
```
$ ./run x -b 2M -m 960
INFO: scrcpy 1.16 <https://github.com/Genymobile/scrcpy>
x/server/scrcpy-server: 1 file pushed. 1.9 MB/s (33634 bytes in 0.017s)
[server] INFO: Device: Xiaomi MI MAX 2 (Android 7.1.1)
INFO: Renderer: opengl
INFO: OpenGL version: 3.0 Mesa 20.1.5
INFO: Trilinear filtering enabled
INFO: Initial texture: 544x960
```
That's all. Scrcpy receives e.g. meta-key combinations and logs them but there is no reaction of the device.

```
[server] INFO: Device rotation requested: landscape
```
Nothing happens.

**Comment by rom1v** [comment]
https://github.com/Genymobile/scrcpy/blob/master/FAQ.md#mouse-and-keyboard-do-not-work

**Comment by eku** [comment]
Yes, I know the FAQ and this point. The setting also exists in my phone but cannot be changed for unknown reason.
Please excuse the disturbance by me.

**Comment by Rudrarokaya** [comment]
okay, after two hours of trying i finally found out the solution. just  run this command in terminal 
(sudo apt install adb ffmpeg libsdl2-2.0-0 make gcc pkg-config meson ninja-build libavcodec-dev libavformat-dev libavutil-dev libsdl2-dev) and after that run scrcpy -m1920 in terminal.
enjoy.

**Comment by Rudrarokaya** [comment]
and for the desktop entry in linux (ubuntu) just create a  file under /usr/share/applications/scrcpy.desktop
[Desktop Entry]
Version=1.0
Type=Application
Name=scrcpy
GenericName=scrcpy
Comment=Screen mirroring application
Exec=scrcpy -m1920%F
Icon=phone-symbolic
Terminal=false
X-MultipleArgs=false
Categories=Development;GTK;
StartupNotify=true

**Comment by Britford** [comment]
In case this helps troubleshoot the display size issues:
tl;dr
scrcpy cuts off part of the screen instead of scaling it.  1200x1920 gets cut down to 1080x1900.  -m 1700 fixes it.

My Lenovo tablet has a 1920x1200 screen.  scrcpy outputs 1920x1080 and the other 120 pixels are just cut off.  In portrait mode, the right side of the screen is cut off.  In landscape, the bottom of the screen is missing.  Touch events do not work.

scrcpy -m 1920   # does not help
scrcpy -m 1700   # seems to work, entire screen is displayed, touch works
I don't know the exact point, but values over 1700 work, less than 1700 or so do not.  1700 seems like an odd value to fix the issue.  Maybe the anomaly will help diagnose the root cause. 

>scrcpy
scrcpy 1.24 <https://github.com/Genymobile/scrcpy>
scrcpy-server: 1 file pushed, 0 skipped. 145.9 MB/s (41159 bytes in 0.000s)
[server] INFO: Device: LENOVO Lenovo TB-X606F (Android 10)
INFO: Renderer: direct3d
INFO: Initial texture: 1200x1920
INFO: New texture: 1080x1920
[server] WARN: Ignore touch event, it was generated for a different device size

<img width="323" alt="image" src="https://user-images.githubusercontent.com/11218574/187548811-03f070f7-8795-4092-80b4-70f97ef77158.png">

<img width="543" alt="2022-08-30_14h44_26" src="https://user-images.githubusercontent.com/11218574/187549294-0d34d404-3b76-4dad-be1a-04c360becd10.png">

**Comment by hhhaiai** [comment]
```
scrcpy 1.24 <https://github.com/Genymobile/scrcpy>
/opt/local/share/scrcpy/scrcpy-server:...ped. 46.3 MB/s (41159 bytes in 0.001s)
[server] INFO: Device: OPPO PACM00 (Android 10)
2024-03-15 16:30:17.679 scrcpy[9815:243314] INFO: Renderer: metal
2024-03-15 16:30:17.686 scrcpy[9815:243314] ERROR: Could not open image codec: /opt/local/share/icons/hicolor/256x256/apps/scrcpy.png
2024-03-15 16:30:17.686 scrcpy[9815:243314] WARN: Could not load icon
2024-03-15 16:30:17.686 scrcpy[9815:243314] INFO: Initial texture: 1080x2280
2024-03-15 16:30:18.087 scrcpy[9815:243314] INFO: New texture: 910x1920
[server] WARN: Ignore touch event, it was generated for a different device size
[server] WARN: Ignore touch event, it was generated for a different device size
[server] WARN: Ignore touch event, it was generated for a different device size
[server] WARN: Ignore touch event, it was generated for a different device size
```

**Comment by rom1v** [comment]
@hhhaiai see #1645/

**Comment by rom1v** [comment]
@hhhaiai And also:

> scrcpy 1.24

Ultra-old. You should upgrade to the latest version.

**Comment by hhhaiai** [comment]
> @hhhaiai And also:
> 
> > scrcpy 1.24
> 
> Ultra-old. You should upgrade to the latest version.

ok

**Comment by hhhaiai** [comment]
> @hhhaiai And also:
> 
> > scrcpy 1.24
> 
> Ultra-old. You should upgrade to the latest version.

already update 2.4, still has same exception

**Comment by rom1v** [comment]
@hhhaiai #1645

```
scrcpy -m1920
```

**Comment by hhhaiai** [comment]
> scrcpy -m1920

It's ok。  why not auto support  general size.

3Q~~

**Comment by hhhaiai** [comment]
> @hhhaiai #1645
> 
> ```
> scrcpy -m1920
> ```
very 3Q. this play is OK .  "scrcpy -m1920"  ,the oppo phone size "INFO: Texture: 912x1920"

check size and update

**Comment by rom1v** [comment]
> why not auto support general size.

Because this is a bug in the encoder: we request a size, it produces a video with a different size without notice (and we can't know that easily on the device side without parsing the encoded video stream).

In general, scrcpy tries a size, then attempts a smaller size if it fails… But some encoders don't fail, but decide to produce a video with a different size instead.

**Comment by rom1v** [comment]
The actual resolution is detected on the client side, once the video stream is parsed/decoded.

The server side thinks the resolution it requested.

When then client sends a mouse events, the server detects a mismatch (which can happen in expected conditions like when the device is physically rotated between the instant an event is triggered on the client side and the time it is injected on the server side), hence the error.

**Comment by hhhaiai** [comment]
> The actual resolution is detected on the client side, once the video stream is parsed/decoded.
> 
> The server side thinks the resolution is what it requested.
> 
> When then client sends a mouse events, the server detects a mismatch (which can happen in expected conditions like when the device is physically rotated between the instant an event is triggered on the client side and the time it is injected on the server side), and ignores the event (with a warning).

How is this solution (* * scrcpy - m1920 * *) implemented?
When the user types the * * scrcpy * * command, they do not directly start the operation. Instead, they first check the resolution and obtain the truly operable resolution (* * INFO: Texture: 910x1920 * *). At this time, they work according to the implementation of the scheme (* * scrcpy - m1920 * *).
Equivalent to, after default detection, we input different resolutions to different devices, and 'scrcpy' is approximately equal to the dynamic parsing resolution, and then execute 'scrcpy - m1920'`

**Comment by hhhaiai** [comment]
> > Instead, they first check the resolution and obtain the truly operable resolution
> 
> That's a lot of work / round trip / guess to workaround an encoder bug which impacts only few devices.
> 
> Btw, in some cases -m1920 does not fix the problem, because many encoders require both dimensions to be a multiple of 8 (for example see [#1645 (comment)](https://github.com/Genymobile/scrcpy/issues/1645#issuecomment-1326684192)). And you would have to handle this encoder bug on rotation too (the encoder might accept the initial size, but after a rotation the width or height exceeds its maximum and sends a video with another size). Anyway, a good solution would be to make the encoder fail if it does not support the requested size, but that's up to the encoders implementations.
> 
> Btw, you could try with another [encoder](https://github.com/Genymobile/scrcpy/blob/master/doc/video.md#encoder).

What you said is indeed an issue, and compatibility issues with a few devices are indeed a troublesome matter
Replace the encoder as per your proposal
```Shell
$scrcpy -- list encoders
Scrcpy 2.4< https://github.com/Genymobile/scrcpy >
INFO: ADB device found:
INFO: -->(USB) PVRKRAAV8IBININ device PACM00
/Usr/local/Cellar/scrcpy/2.4/share/scrcpy/scrcpy server: 1 file pushed, 0 skipped 282.3 MB/s (69007 bytes in 0.000s)
[server] INFO: Device: [OPPO] OPPO PACM00 (Android 10)
[server] INFO: List of video encoders:
--Video code=h264-- video encoder='OMX MTK VIDEO ENCODER AVC '
--Video code=h264-- video encoder='c2. android. avc. encoder '
--Video code=h264-- video encoder='OMX. Google. h264. encoder '
[server] INFO: List of audio encoders:
--Audio code=opus -- audio encoder='c2. android. opus. encoder '
--Audio code=aac -- audio encoder='OMX. Google. aac. encoder '
--Audio code=aac -- audio encoder='c2. android. aac. encoder '
--Audio code=flac -- audio encoder='OMX. Google. flac. encoder '
--Audio code=flac -- audio encoder='c2. android. flac. encoder '
```
There are two decoders that can work. Can the decoders be dynamically switched during operation
```
#Success
--Video code=h264-- video encoder='c2. android. avc. encoder '
--Video code=h264-- video encoder='OMX. Google. h264. encoder '
```

**Comment by rom1v** [comment]
> it is expected that the number of devices with this issue will continue to increase.

Why that? If the encoder does not support the requested size, it should fail (as most of encoders do).

Also, another problem is that the encoder produces a video at a different resolution than that requested, but we don't know how the video is scaled/cropped (some may add some black borders, other may stretch the video…), so we don't know for sure how to map input events back to thebdevice coordinates. Btw, in your case, one dimension is 910, which is not even a multiple of 8.

> Can the decoders be dynamically switched during operation

No, one must be explicitly selected. There is no absolute criteria to select "the best".

**Comment by hhhaiai** [comment]
> > it is expected that the number of devices with this issue will continue to increase.
> 
> Why that? If the encoder does not support the requested size, it should fail (as most of encoders do).
> 
> Also, another problem is that the encoder produces a video at a different resolution than that requested, but we don't know how the video is scaled/cropped (some may add some black borders, other may stretch the video…), so we don't know for sure how to map input events back to thebdevice coordinates. Btw, in your case, one dimension is 910, which is not even a multiple of 8.
> 
> > Can the decoders be dynamically switched during operation
> 
> No, one must be explicitly selected. There is no absolute criteria to select "the best".

ok.  3Q

**Comment by riotcult** [comment]
> [#1518 (comment)](https://github.com/Genymobile/scrcpy/issues/1518#issuecomment-655499111)
> 
> > But probably not related.
> 
> Oh in fact, that might be related: the client sends clicks for a window having dimensions different from the ones provided by the device.
> 
> If you remove these lines, it should "work" (it may click at a wrong location though):
> 
> https://github.com/Genymobile/scrcpy/blob/e99b896ae2b6a59d5f7f2b76a3dd9c29f4838013/server/src/main/java/com/genymobile/scrcpy/Device.java#L134-L138
> 
> > ```
> > INFO: Initial texture: 1080x2336
> > INFO: New texture: 888x1920
> > ```
> 
> What if you run `scrcpy -m1920`? Do clicks work in that case?

Hey man I am new to all this can you guide me through all that I have to do for this to work?

**Comment by rom1v** [comment]
@riotcult Read https://github.com/Genymobile/scrcpy/blob/master/doc/windows.md#run
Execute `scrcpy -m1920`.
